### ISSUE TITLE
Behavior policies

### ISSUE THREAD

**Issue by sztomi** [conan-io/conan#3358]
CMake preserves backward compatibility by introducing policies that control applying  breaking behaviors or sticking with the old way.

For example, fixing https://github.com/conan-io/conan/issues/3303 is a breaking change (applying environment variables where previously they weren't). A policy could control whether or not to apply the environment.

In general, this would allow introducing breaking changes with ease. Riskier features could be rolled out in an opt-in fashion first. Or when a bug is fixed, they could be opt-out, still allowing users who rely on the old behavior to use it.

In our use-case, setting the policies in the conan config would be enough. We wouldn't need something like setting policies in a conanfile.

Implementation-wise, I think refactoring the parts set by policies into callable objects would be the way to go. `__call__` would select the implementation based on the policy value.

```py
class PolicyBehavior(ABC):
  def __init__(self, key):
    self.key = key

  def __call__(*args, **kwargs):
    if config.get_policy(self.key) == "OLD":
      return self.run_old(*args, **kwargs)
    else:
      return self.run_new(*args, **kwargs)

  @abstractmethod
  def run_old(*args, **kwargs):
    pass

  @abstractmethod
  def run_new(*args, **kwargs):
    pass
```

---

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by lasote** [comment]
The bad part is the maintenance of the code base including the test suite complexity. I think we probably should avoid it as much as possible.

**Comment by sztomi** [comment]
I agree that it should be used sparingly, but it can be a valueable tool.

**Comment by memsharded** [comment]
Closing this ticket as not planned at the moment, so far the breaking behaviors have been introduced via opt-ins, seems good enough for the resources we have (policy based behavior would require more effort)
